exonum/src/runtime/dispatcher/mod.rs

```diff
@@ -409,23 +593,147 @@ impl Dispatcher {
     }
 
     /// Commits to service instances and artifacts marked as pending in the provided `fork`.
-    pub(crate) fn commit_block(&mut self, fork: Fork) -> Patch {
+    pub(crate) fn commit_block(&mut self, mut fork: Fork) -> Patch {
         let mut schema = Schema::new(&fork);
         let pending_artifacts = schema.take_pending_artifacts();
         let modified_instances = schema.take_modified_instances();
+
+        // Blocks until all migrations are completed with the expected outcome.
+        // The node will panic if the local outcome of a migration is unexpected.
+        for (state, _) in &modified_instances {
+            let migration_hash = state
+                .status
+                .as_ref()
+                .and_then(InstanceStatus::completed_migration_hash);
+            if let Some(hash) = migration_hash {
+                let instance_name = &state.spec.name;
+                let local_result = schema.local_migration_result(instance_name);
+                let local_result = self.block_on_migration(instance_name, hash, local_result);
+                schema.add_local_migration_result(instance_name, local_result);
+            }
+        }
+
+        // Rollback migrations.
+        let migration_rollbacks =
+            modified_instances
+                .iter()
+                .filter_map(|(state, prev_status)| match (&state.status, prev_status) {
+                    (Some(InstanceStatus::Stopped), Some(InstanceStatus::Migrating { .. })) => {
```

Comment:
It isn't really obvious why this combinanion means that migration should be rolled back.